manticore/core/executor.py

```diff
@@ -48,29 +48,84 @@ def newFunction(self, *args, **kw):
-        pass
+    @property        
+    def executor(self):
+        return self._executor
 
-    def priority(self, state_id):
-        ''' A numeric value representing likelihood to reach the interesting program spot '''
-        return 1.0
+    @contextmanager
+    def locked_context(self, key=None, default=dict):
+        ''' Policy shared context dictionary '''
+        keys = ['policy']
+        if key is not None:
+            keys.append(key)
+        with self._executor.locked_context('.'.join(keys), default) as policy_context:
+            yield policy_context
+
+    def _add_state_callback(self, state_id, state):
+        ''' Save sumarize(state) on policy shared context before 
+            the state is stored
+        '''
+        summary = self.summarize(state)
+        if summary is None:
+            return
+        with self.locked_context('summaries', dict) as ctx:
+            ctx[state_id] = summary
+
+    def summarize(self, state):
+        '''
+            Extract the relevant information from a state for later
+            prioritization
+        '''
+        return None
+
+    def choice(self, state_ids):
+        ''' Select a state id from state_ids.
+            self.context has a dict mapping state_ids -> summarize(state)'''
+        raise NotImplemented
 
 class Random(Policy):
-    def __init__(self):
-        super(Random, self).__init__()
+    def __init__(self, executor, *args, **kwargs):
+        super(Random, self).__init__(executor, *args, **kwargs)
+
+    def choice(self, state_ids):
+        return random.choice(state_ids)
 
-    def features(self, state):
-        ''' Save state features for prioritization before a state is stored '''
-        pass
+class Uncovered(Policy):
+    def __init__(self, executor, *args, **kwargs):
+        super(Uncovered, self).__init__(executor, *args, **kwargs)
+        #hook on the necesary executor signals 
+        #on callbacks save data in executor.context['policy']
 
-    def priority(self, state_id):
-        ''' A numeric value representing likelihood to reach the interesting program spot '''
-        return 1.0
+        self._executor.will_load_state += self._register
 
+    def _register(self, *args):
+        self.executor.will_execute_instruction += self._visited_callback
+
+    def _visited_callback(self, state, instr):
+        ''' Maintain our own copy of the visited set
+        '''
+        pc = state.platform.current.PC
+        with self.locked_context('visited', set) as ctx:
+            ctx.add(pc)
+
+    def summarize(self, state):
+        ''' Save the last pc before storing the state '''
+        return state.cpu.PC
+
+    def choice(self, state_ids):
+        interesting = set()
+        with self.locked_context() as policy_ctx: 
+            visited = policy_ctx.get('visited', set())
+            lastpc = policy_ctx.get('summaries', dict())
+            for _id in state_ids:
+                if lastpc.get(_id, None) not in visited:
```

Comment:
`get(x, None)` is equivalent to just `get(x)` (this happens in a number of places)